lib/pyld/iri_resolver.py

```diff
@@ -204,4 +208,77 @@ def resolve(relative_iri: str, base_iri: str = None) -> str:
     relative_iri = base_path + relative_iri
     relative_iri = remove_dot_segments(relative_iri)
 
-    return base_iri[:base_slash_after_colon_pos] + relative_iri
\ No newline at end of file
+    return base_iri[:base_slash_after_colon_pos] + relative_iri
+
+def unresolve(absolute_iri: str, base_iri: str = ""):
+    """
+    Unresolves a given absolute IRI to an IRI relative to the given base IRI.
+
+    :param base: the base IRI.
+    :param iri: the absolute IRI.
+
+    :return: the relative IRI if relative to base, otherwise the absolute IRI.
+    """
+    # skip IRI processing
+    if not base_iri:
+        return absolute_iri
+
+    base = urlparse(base_iri)
+
+    if not base.scheme:
+        raise ValueError(f"Found invalid baseIRI '{base_iri}' for value '{absolute_iri}'")
+
+    # compute authority (netloc) and strip default ports
+    base_authority = parse_authority(base)
+
+    rel = urlparse(absolute_iri)
+    # compute authority (netloc) and strip default ports
+    rel_authority = parse_authority(rel)
+
+    # schemes and network locations (authorities) don't match, don't alter IRI
+    if not (base.scheme == rel.scheme and base_authority == rel_authority):
+        return absolute_iri
+
+    # remove path segments that match (do not remove last segment unless there
+    # is a hash or query
+    base_segments = remove_dot_segments(base.path).split('/')
+    iri_segments = remove_dot_segments(rel.path).split('/')
+    last = 0 if (rel.fragment or rel.query) else 1
+    while (len(base_segments) and len(iri_segments) > last and
+            base_segments[0] == iri_segments[0]):
+        base_segments.pop(0)
+        iri_segments.pop(0)
+
+    # use '../' for each non-matching base segment
+    rval = ''
+    if len(base_segments):
+        # don't count the last segment (if it ends with '/' last path doesn't
+        # count and if it doesn't end with '/' it isn't a path)
+        base_segments.pop()
+        rval += '../' * len(base_segments)
+
+    # prepend remaining segments
+    rval += '/'.join(iri_segments)
+
+    # build relative IRI using urlunparse with empty scheme/netloc
+    return urlunparse(('', '', rval, '', rel.query or '', rel.fragment or '')) or './'
+
+def parse_authority(parsed_iri) -> str:
```

Comment:
- Perhaps a type annotation for `parsed_iri` would be handy;
- The terms of IRI, URI, and URL tend to cause confusion; I have heard somewhere that there are groups in W3C arguing towards using URL for everything. Would be interested to hear your thoughts.